### ISSUE TITLE
Feature Link to Story is missing

### ISSUE THREAD

**Issue by gmcgee2014** [solidify/jira-azuredevops-migrator#557]
**Describe the bug**
We are unable to preserve the link between the Feature and Story in our migrations. We've tried various combinations of the link map settings with no luck. We are seeing parent / child relationships created for Story to Task and for Bug to Task. Need help with solving the link between Feature and Story. We're running jira-azuredevops migrator 2.3.117

**To Reproduce**
Here's the json config we're currently using:

{
  "source-project": "OurLegacyJiraProject",
  "target-project": "OurNewAzDoProject",
  "query": "((issuekey = ITSM-41564) or (cf[10006] = ITSM-41564)) order by issuekey ASC",
  "using-jira-cloud": false,
  "workspace": "C:\\Temp\\JiraExport\\",
  "epic-link-field": "Epic Link",
  "sprint-field": "Sprint",
  "download-options": 7,
  "batch-size": 20,
  "log-level": "Info",
  "attachment-folder": "Attachments",
  "user-mapping-file": "users.txt",
  "base-area-path": "Migrated",  
  "base-iteration-path": "Migrated",  
  "ignore-failed-links": false,
  "process-template": "Agile",
  "link-map": {
    "link": [
     {
      "source": "Epic Link",
      "target": "System.LinkTypes.Hierarchy-Reverse"
      },
     {
      "source": "Epic",
      "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
      "source": "Parent",
      "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
      "source": "Child",
      "target": "System.LinkTypes.Hierarchy-Forward"
      }, 
      {
      "source": "is blocked by",
      "target": "System.LinkTypes.Hierarchy-Forward"
      }, 
      {
      "source": "Relates",
      "target": "System.LinkTypes.Related"
      },
      {
      "source": "Duplicate",
      "target": "System.LinkTypes.Duplicate-Forward"
      },
      {
      "source": "Dependency",
      "target": "System.LinkTypes.Dependency-Forward"
      },
      {
      "source": "Problem/Incident",
      "target": "System.LinkTypes.Related"
      },
      {
      "source": "depends on",
      "target": "System.LinkTypes.Dependency-Forward"
      },
      {
      "source": "duplicates",
      "target": "System.LinkTypes.Duplicate-Forward"
      },
      {
      "source": "causes",
      "target": "System.LinkTypes.Related-Forward"
      },
      {
      "source": "relates to",
      "target": "System.LinkTypes.Related"
      },
      {
      "source": "is depended on by",
      "target": "System.LinkTypes.Dependency-Reverse"
      },
      {
      "source": "is blocked by",
      "target": "System.LinkTypes.Dependency-Forward"
      }, 
      {
      "source": "is duplicated by",
      "target": "System.LinkTypes.Duplicate-Reverse"
      },
      {
      "source": "is caused by",
      "target": "System.LinkTypes.Related-Reverse"
      }
    ]
  },
  "type-map": {
    "type": [
      {
        "source": "Epic",
        "target": "Epic"
      },
      {
        "source": "Feature",
        "target": "Feature"
      },	
      {
        "source": "Story",
        "target": "User Story"
      },
      {
        "source": "Bug",
        "target": "Bug"
      },
      {
        "source": "Task",
        "target": "Task"
      },
      {
        "source": "Sub-task",
        "target": "Task"
      }
    ]
  },
  "field-map": {
    "field": [
      {
        "source": "summary",
        "target": "System.Title",
        "mapper": "MapTitle"
      },
      {
        "source": "assignee",
        "target": "System.AssignedTo",
        "mapper": "MapUser"
      },
      {
        "source": "description",
        "target": "System.Description",
        "mapper": "MapRendered"
      },
      {
        "source": "Acceptance Criteria",
        "source-type": "name",
	"target": "Microsoft.VSTS.Common.AcceptanceCriteria"
      },
      {
        "source": "priority",
        "target": "Microsoft.VSTS.Common.Priority",
        "mapping": {
          "values": [
            {
              "source": "Highest",
              "target": "1"
            },
            {
              "source": "High",
              "target": "2"
            },
            {
              "source": "Medium",
              "target": "3"
            },
            {
              "source": "Low",
              "target": "3"
            },
            {
              "source": "Lowest",
              "target": "4"
            }
          ]
        }
      },
      {
        "source": "labels",
        "target": "System.Tags",
        "mapper": "MapTags"
      },
      {
        "source": "comment",
        "target": "System.History",
        "mapper": "MapRendered"
      },
      {
        "source": "status",
        "target": "System.State",
        "for": "User Story",
        "mapping": {
          "values": [
            {
              "source": "Initiate",
              "target": "Backlog"
            },
            {
              "source": "Requirements",
              "target": "Analyze"
            },
            {
              "source": "Design & Build",
              "target": "Build"
            },
            {
              "source": "Build/UAT",
              "target": "Integrate-Test"
            },
            {
              "source": "Deploy",
              "target": "Accepted"
            },
            {
              "source": "Monitor",
              "target": "Integrate-Test"
            },
            {
              "source": "Blocked/On Hold",
              "target": "Blocked"
            },
            {
              "source": "Completed",
              "target": "Accepted"
            }
          ]
        }
      },
      {
        "source": "status",
        "target": "System.State",
        "for": "Feature",
        "mapping": {
          "values": [
            {
              "source": "Initiate",
              "target": "Funnel"
            },
            {
              "source": "Requirements",
              "target": "Analyzing"
            },
            {
              "source": "Design & Build",
              "target": "Implementing"  //Implementing - EDS
            },
            {
              "source": "Build/UAT",
              "target": "Validating on staging"
            },
            {
              "source": "Deploy",
              "target": "Deploying to Production"
            },
            {
              "source": "Blocked/On Hold",
              "target": "Blocked"
            },
            {
              "source": "Monitor",
              "target": "Releasing"
            },
            {
              "source": "Completed",
              "target": "Done"
            }
          ]
        }
      },
      {
        "source": "status",
        "target": "System.State",
        "for": "Bug",
        "mapping": {
          "values": [
            {
              "source": "Initiate",
              "target": "Backlog"
            },
            {
              "source": "Requirements",
              "target": "Analyze"
            },
            {
              "source": "Design & Build",
              "target": "Build"
            },
            {
              "source": "Build/UAT",
              "target": "Integrate-Test"
            },
            {
              "source": "Blocked/On Hold",
              "target": "Blocked"
            },
            {
              "source": "Deploy",
              "target": "Accepted"
            },
{
              "source": "Monitor",
              "target": "Accepted"
            },
            {
              "source": "Completed",
              "target": "Accepted"
            }
          ]
        }
      },	
      {
        "source": "status",
        "target": "System.State",
        "for": "Task",
        "mapping": {
          "values": [
            {
              "source": "To Do",
              "target": "New"
            },
            {
              "source": "In Progress",
              "target": "Active"
            },
	    {
              "source": "On Hold",
              "target": "Blocked"
            },
            {
              "source": "Testing",
              "target": "Active"
            },
            {
              "source": "Closed",
              "target": "Closed"
            }
          ]
        }
      },
      {
        "source": "Story Points",
        "source-type": "name",
        "target": "Microsoft.VSTS.Scheduling.StoryPoints",
        "not-for": "Task"
      },
      {
        "source": "timeestimate",
        "target": "Microsoft.VSTS.Scheduling.RemainingWork",
        "mapper": "MapRemainingWork",
        "for": "Bug,Task"
      },
      {
        "source": "description",
        "target": "Microsoft.VSTS.TCM.ReproSteps",
        "for": "Bug"
      },
      {
        "source": "environment",
        "source-type": "name",
        "target": "Microsoft.VSTS.TCM.SystemInfo",
        "for": "Bug,Epic"
      }
    ]
  }
}

**Expected behavior**
Once we complete the export and import into AzDo we should see a link in the Related links section of the AzDo work item (i.e. Feature and Story).

**Screenshots**
Here's the view from the feature in Jira of the child stories:
![image](https://user-images.githubusercontent.com/114501288/192571176-4a1c6e43-7967-4796-9fcb-02f44087d29b.png)

Here's the view from the imported Feature in Azdo
![image](https://user-images.githubusercontent.com/114501288/192571358-06c664b0-5779-4c75-b87f-a76b4c491c18.png)

**Comment by Alexander-Hjelm** [comment]
Does the error still persist with the latest prerelease?
[jira-azuredevops-migrator-3.0.15.zip](https://github.com/solidify/jira-azuredevops-migrator/files/9662494/jira-azuredevops-migrator-3.0.15.zip)

We have fixed a lot of bugs since 2.3.117, so it might be that the issue has been resolved already. Try the prerelease build and see if it works for you, otherwise I'll try to reproduce.

Note that we still have an issue with the prerelease where attachments are sometimes duplicated. See: https://github.com/solidify/jira-azuredevops-migrator/issues/556

**Comment by Alexander-Hjelm** [comment]
Looks like it is the AddRemoveSingleLink() method that complains about the fieldparameter:

https://github.com/solidify/jira-azuredevops-migrator/blob/58f1518fc6a8da59cd31a36018f176ed9a1499ba/src/WorkItemMigrator/JiraExport/RevisionUtils/LinkMapperUtils.cs#L44

And the only invocation where field is nullable is for the epic links:

https://github.com/solidify/jira-azuredevops-migrator/blob/58f1518fc6a8da59cd31a36018f176ed9a1499ba/src/WorkItemMigrator/JiraExport/JiraMapper.cs#L216

What is EpicLinkField set to in your config?

**Comment by gmcgee2014** [comment]
We actually don't have epics in our Jira project that we care about migrating. I've tried removing that settings from the config-agile.json. But still getting that error message. Is that setting required still?

**Comment by Alexander-Hjelm** [comment]
It could be that it's needed anyway, not sure what it defaults to. Try!

**Comment by gmcgee2014** [comment]
Even though we don't have any epics in the project, I've tried adding in epic mapping for states. Here's the current config file with epic info. However I'm still getting the same error. 

 "epic-link-field": "Epic Link",
  "sprint-field": "Sprint",
  "download-options": 7,
  "batch-size": 20,
  "log-level": "Info",
  "attachment-folder": "Attachments",
  "user-mapping-file": "users.txt",
  "base-area-path": "ServiceNow ART/IT Service Management",  //We need to change this, ex: TargetProcessPOC\Integration ART\Greg Team
  "base-iteration-path": "Migrated",  //We need to change this
  "ignore-failed-links": true,
  "process-template": "Agile",
  "link-map": {
    "link": [
      {
      "source": "Epic",
      "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
      "source": "Parent",
      "target": "System.LinkTypes.Hierarchy-Reverse"
      },
      {
      "source": "Child",
      "target": "System.LinkTypes.Hierarchy-Forward"
      }, 
      {
      "source": "is blocked by",
      "target": "System.LinkTypes.Hierarchy-Forward"
      }, 
      {
      "source": "Relates",
      "target": "System.LinkTypes.Related"
      },
      {
      "source": "Duplicate",
      "target": "System.LinkTypes.Duplicate-Forward"
      },
      {
      "source": "Dependency",
      "target": "System.LinkTypes.Dependency-Forward"
      },
      {
      "source": "Problem/Incident",
      "target": "System.LinkTypes.Related"
      },
      {
      "source": "depends on",
      "target": "System.LinkTypes.Dependency-Forward"
      },
      {
      "source": "duplicates",
      "target": "System.LinkTypes.Duplicate-Forward"
      },
      {
      "source": "causes",
      "target": "System.LinkTypes.Related-Forward"
      },
      {
      "source": "relates to",
      "target": "System.LinkTypes.Related"
      },
      {
      "source": "is depended on by",
      "target": "System.LinkTypes.Dependency-Reverse"
      },
      {
      "source": "is blocked by",
      "target": "System.LinkTypes.Dependency-Forward"
      }, 
      {
      "source": "is duplicated by",
      "target": "System.LinkTypes.Duplicate-Reverse"
      },
      {
      "source": "is caused by",
      "target": "System.LinkTypes.Related-Reverse"
      }
    ]
  },
  "type-map": {
    "type": [
      {
        "source": "Epic",
        "target": "Epic"
      },
      {
        "source": "Feature",
        "target": "Feature"
      },	
      {
        "source": "Story",
        "target": "User Story"
      },
      {
        "source": "Bug",
        "target": "Bug"
      },
      {
        "source": "Task",
        "target": "Task"
      },
      {
        "source": "Sub-task",
        "target": "Task"
      }
    ]
  },
  "field-map": {
    "field": [
      {
        "source": "summary",
        "target": "System.Title",
        "mapper": "MapTitle"
      },
      {
        "source": "assignee",
        "target": "System.AssignedTo",
        "mapper": "MapUser"
      },
      {
        "source": "description",
        "target": "System.Description",
        "mapper": "MapRendered"
      },
      {
        "source": "Acceptance Criteria",
        "source-type": "name",
	"target": "Microsoft.VSTS.Common.AcceptanceCriteria",
        "for": "Feature,User Story" 
      },
      {
        "source": "priority",
        "target": "Microsoft.VSTS.Common.Priority",
        "mapping": {
          "values": [
            {
              "source": "Highest",
              "target": "1"
            },
            {
              "source": "High",
              "target": "2"
            },
            {
              "source": "Medium",
              "target": "3"
            },
            {
              "source": "Low",
              "target": "3"
            },
            {
              "source": "Lowest",
              "target": "4"
            }
          ]
        }
      },
      {
        "source": "labels",
        "target": "System.Tags",
        "mapper": "MapTags"
      },
      {
        "source": "comment",
        "target": "System.History",
        "mapper": "MapRendered"
      },
      {
        "source": "status",
        "target": "System.State",
        "for": "Epic",
        "mapping": {
          "values": [
            {
              "source": "Funnel",
              "target": "Funnel"
            },
            {
              "source": "Reviewing",
              "target": "Reviewing"
            },
            {
              "source": "Analyzing",
              "target": "Analyzing"
            },
            {
              "source": "Portfolio Backlog",
              "target": "Backlog"
            },
            {
              "source": "Implementing",
              "target": "Implementing - MVP"
            },
            {
              "source": "Done",
              "target": "Done"
            }
          ]
        }
      },
      {
        "source": "status",
        "target": "System.State",
        "for": "User Story",
        "mapping": {
          "values": [
            {
              "source": "Initiate",
              "target": "Backlog"
            },
            {
              "source": "Requirements",
              "target": "Analyze"
            },
            {
              "source": "Design & Build",
              "target": "Build"
            },
            {
              "source": "Build/UAT",
              "target": "Integrate-Test"
            },
            {
              "source": "Deploy",
              "target": "Accepted"
            },
            {
              "source": "Monitor",
              "target": "Integrate-Test"
            },
            {
              "source": "Blocked/On Hold",
              "target": "Blocked"
            },
            {
              "source": "Completed",
              "target": "Accepted"
            }
          ]
        }
      },
      {
        "source": "status",
        "target": "System.State",
        "for": "Feature",
        "mapping": {
          "values": [
            {
              "source": "Initiate",
              "target": "Funnel"
            },
            {
              "source": "Requirements",
              "target": "Analyzing"
            },
            {
              "source": "Design & Build",
              "target": "Implementing"  
            },
            {
              "source": "Build/UAT",
              "target": "Validating on staging"
            },
            {
              "source": "Deploy",
              "target": "Deploying to Production"
            },
            {
              "source": "Blocked/On Hold",
              "target": "Blocked"
            },
            {
              "source": "Monitor",
              "target": "Releasing"
            },
            {
              "source": "Completed",
              "target": "Done"
            }
          ]
        }
      },
      {
        "source": "status",
        "target": "System.State",
        "for": "Bug",
        "mapping": {
          "values": [
            {
              "source": "Initiate",
              "target": "Backlog"
            },
            {
              "source": "Requirements",
              "target": "Analyze"
            },
            {
              "source": "Design & Build",
              "target": "Build"
            },
            {
              "source": "Build/UAT",
              "target": "Integrate-Test"
            },
            {
              "source": "Blocked/On Hold",
              "target": "Blocked"
            },
            {
              "source": "Deploy",
              "target": "Accepted"
            },
{
              "source": "Monitor",
              "target": "Accepted"
            },
            {
              "source": "Completed",
              "target": "Accepted"
            }
          ]
        }
      },	
      {
        "source": "status",
        "target": "System.State",
        "for": "Task",
        "mapping": {
          "values": [
            {
              "source": "To Do",
              "target": "New"
            },
            {
              "source": "In Progress",
              "target": "Active"
            },
	    {
              "source": "On Hold",
              "target": "Blocked"
            },
            {
              "source": "Testing",
              "target": "Active"
            },
            {
              "source": "Closed",
              "target": "Closed"
            }
          ]
        }
      },
      {
        "source": "Story Points",
        "source-type": "name",
        "target": "Microsoft.VSTS.Scheduling.StoryPoints",
        "not-for": "Task"
      },
      {
        "source": "timeestimate",
        "target": "Microsoft.VSTS.Scheduling.RemainingWork",
        "mapper": "MapRemainingWork",
        "for": "Bug,Task"
      },
      {
        "source": "description",
        "target": "Microsoft.VSTS.TCM.ReproSteps",
        "for": "Bug"
      },
      {
        "source": "environment",
        "source-type": "name",
        "target": "Microsoft.VSTS.TCM.SystemInfo",
        "for": "Bug,Epic"
      }
    ]
  }
}

![image](https://user-images.githubusercontent.com/114501288/193641628-d9266947-767d-4c6a-bc1e-4c50888ccd13.png)
